src/helpers/cloneChildren.js

```diff
@@ -1,38 +1,113 @@
+function createRecursiveRule(rules) {
+    return {
+        predicate: () => true,
+        clone: (child, childNames) => {
+            return React.cloneElement(child, {},
+                cloneChildren(rules, child.props && child.props.children, childNames));
+        }
     }
+}
 
-    return React.Children.map(children, (child) => {
-        if (typeof child !== 'object' || child === null) {
-            return child;
+/**
+ * A common function for cloning Errors element that takes care of injecting
+ * required error data
+ *
+ * @param {Object} errors aaaaa
+ * @param {Object} fieldErrors bbbb
+ * @return {Object} rule for cloning Errors element
+ */
+export function createErrorsRule({ errors = [], fieldErrors = {} }) {
+    return {
+        predicate: child => child.type && child.type.displayName === 'Errors',
+        clone: child => {
+            return React.cloneElement(
+                child,
+                {
+                    errors: errors,
+                    fieldErrors: fieldErrors
+                },
+                child.props && child.props.children
+            );
         }
+    }
+}
+
+/*
+ * Get extra properties for something we are going to weave our formable magic into.
+ */
+function getFormableComponentProperties(errors, fieldErrors, onSubmit, onChange) {
+    return (child, childNames) => {
+        warning(!child.ref, `Attempting to attach ref "${child.ref}" to "${child.props.name}" will be bad for your health`);
+        warning(childNames.indexOf(child.props.name) === -1, `Duplicate name "${child.props.name}" found. Duplicate fields will be ignored`);
+        childNames.push(child.props.name);
+
+        return {
+            ref: child.ref || child.props.name,
+            onChange: compose(onChange, child.props.onChange || identity),
+            onSubmit: compose(onSubmit, child.props.onSubmit || identity),
+            errors: errors,
+            fieldErrors: child.props.fieldErrors || fieldErrors[child.props.name]
+        };
+    }
+}
 
-        if (predicate(child)) {
+/*
+ * Standard cloning rule for something react-formable
+ */
+export function createFormableRule(
+    { errors = [], fieldErrors = {} },
+    onSubmit = identity,
+    onChange = identity) {
+    return {
+        predicate: child => child.props && child.props.name,
+        clone: (child, childNames) => {
             return React.cloneElement(
                 child,
-                getProps(child),
+                getFormableComponentProperties(errors, fieldErrors, onSubmit, onChange)(child, childNames),
                 child.props && child.props.children
             );
         }
+    }
+}
+
+/**
+ * Clones a child subtree using the supplied rules which are composed of predicates
+ * and clone instructions.
+ *
+ * @param  {array} rules used to predicate and clone
+ * @param  {Function} children The children to iterate over
+ * @param  {array=} childNames optionally and ONLY supplied for internal recursion
```

Comment:
`=` here is used to denote optional, not sure which standard we might use for that. The description tries to exaplain why it's optional.